Rewrites/dba array string table

Quix0r opened the pull request:
Rewrite:
- avoid having `array|string` for `$table` to have a strong `string` type-hint for `$table`
- you now have to do it for yourself by giving `schema.table` as parameter

Quix0r committed:
Fixed:
- Uncaught Exception TypeError: "Argument 3 passed to
  Friendica\Protocol\DFRN::createActivity() must be of the type string, null
  given, called ..."

Quix0r committed:
Changes:
- added more type-hints

Quix0r committed:
Changes:
- added type-hints
- added some missing documentation

Quix0r committed:
Removed comment

Quix0r committed:
Rewrite:
- avoid having array|string for $table to have a "string" type-hint for $table
- you now have to do it for yourself by giving 'schema.table' as parameter

Quix0r commented:
**SOLVED**: I have this here: `Table 'test.information_schema.tables' doesn't exist `

Quix0r committed:
WIP: Properly some fixes? Also why is DROP VIEW IF EXISTS is being killed?

Quix0r commented:
**SOLVED**: Also strange is that `DROP VIEW IF EXISTS` is being removed from `database.sql`?

Quix0r commented:
**SOLVED**: In `Database/View::createView()` a call to `self::isView()` might have prevented `DROP VIEW IF EXISTS` being added.

Quix0r commented:
And can we move `isView/Table()` to `Database` class instead? They feel very generic to me.

Quix0r commented:
Okay, I have added a debug logger for the SQL statement being used and I got `NFORMATION_SCHEMA.TABLES` being wrapped into back-ticks.  That needs fixing!

Quix0r committed:
Fixed:
- since $table can be now schema.table, this needs being taken care off
- maybe this foreach(explode()) is a bit an overkill as only 1 or 2 entries
  can be found in it

Quix0r commented:
Okay, I have added a debug logger for the SQL statement being used and I got `NFORMATION_SCHEMA.TABLES` being wrapped into back-ticks.  That needs fixing!

Quix0r commented on src/Database/DBA.php:
```diff
@@ -524,20 +521,20 @@ public static function count($table, array $condition = [], array $params = []):
 		foreach ($tables as $schema => $table) {
 			if (is_numeric($schema)) {
-				$quotedTables[] = self::quoteIdentifier($table);
+				$str = '';
```
Well, some quick'n'dirty code here. Normally only 1 or 2 elements should be expected here.

Quix0r committed:
Changes:
- Database->$driver can no longer be NULL, an empty string is fine anyway

Quix0r commented:
Now getting `Error: Call to a member function rollBack() on null` when local database configuration is not valid. This originates from `DatabaseTestTrait::setUpDb()`. It looks like the code there assumes that the database connection is established.

Quix0r committed:
Changes:
- let's at least log the error message, yes it does flood your logfile

Quix0r committed:
Fixed:
- PHPUnit\Framework\Error\Notice:  Undefined index: language in /.../src/Util/EMailer/MailBuilder.php on line 122

Quix0r committed:
Added simple wrapper script to run PHPUnit tests for you. To use this, you need
to set at least 4 variables in your ~/.bashrc (recommended):

export FRIENDICA_MYSQL_HOST="localhost"
export FRIENDICA_MYSQL_DATABASE="friendica_test"
export FRIENDICA_MYSQL_USER="friendica_test"
export FRIENDICA_MYSQL_PASSWWORD="friendica_test"

Then you can put ./bin/dev/run_tests.sh into your .git/hooks/pre-commit file

Quix0r committed:
Ops, WWORD ...

nupplaphil commented on tests/Util/Database/StaticDatabase.php:
```diff
@@ -179,7 +180,8 @@ public static function statConnect(array $server)
 		} catch (PDOException $e) {
-			/// @TODO At least log exception, don't ignore it!
+			// Log exception
+			Logger::error('Cannot establish database connection', ['exception' => $e, 'host' => $server, 'dbname' => $db]);
```
Plz no dependency injection directly without using the constructor, because that's where the party gets started :)

`Logger::notice()` is calling `self::logger->notice()`, which has to be initialized with proper `LoggerInterface $logger` which is created by a factory using... the `Database` itself :)) and `StaticDatabase` must be called before any DB call, otherwise there may be a transaction mismatching during tests

Quix0r replied:
So I better use `$this->logger` instead?

Quix0r replied:
No, there is no object to use `$this`. :-( Ops, my mistake.

Quix0r replied:
So, how do I get a logger in a static (non-object) method? `Logger::foo()` seem to have worked. At least I got the message logged.

Quix0r replied:
If we could make `Database::$logger` static it would solve this? Plus rewrite all invocations to `self::$logger` of course.

Quix0r replied:
What I have also noticed in that method is that a lot variables are **only** initialized when the first `if()` block is executed. If it is not executed they (e.g. `$db_host`, `$db_user`, `$db_pw`) are not initialized but used later leading maybe to "Undefined variable" errors. If we could initialize all our variables clearly, means on the method start, this can be prevented.

nupplaphil replied:
There is no easy possibility to add a logger here, that's why there isn't any yet and instead a placeholder.. This execution point is a critical state during a testrun, and tbh I'd like to leave here no further logic (yet) because I spent hours debugging cases, where transactions weren't fully closed and strange/unpredictable errors occur (sometimes -mainly during debugging other errors :) ...)

Quix0r replied:
Okay, so you suggest me to revert my changes here?

MrPetovan replied:
Yes please!

nupplaphil commented on bin/dev/run_tests.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
```
Did you give `autotest.sh` in the root path a try? :)

Quix0r replied:
The script `autotest.sh` requires to set environment variables `MYSQL_*` which are okay for a single project development (e.g. being set in `~/.bashrc`) but not for multi-project developments. This is why I had prefixed my variables with `FRIENDICA_` so they are Friendica-specific.

nupplaphil replied:
maybe you cold extend `autorun.sh` beeing more multi-project friendly :-)

tbh I'm not using it very often anymore, so I guess you could rewrite it completely too or drop and replace it ..

Quix0r replied:
Also I would like to move it out of the root directory to `bin/dev/` as this is clearly a developer's script.

Quix0r committed:
Ops, autorun.sh is there but looks like lesser "multi-project" friendly as
MYSQL_* must be defined. I might still use my version as that is
project-specific.

MrPetovan commented:
> * avoid having `array|string` for `$table` to have a strong `string` type-hint for `$table`

I disagree with the "strong". `string` is the widest type-hint you can have for a parameter and as such, `'(schema.)table'` is actually less specific than `'table'|['schema' => 'table']`. 

You can see the issue with `DBA::buildTableString` where you have to perform several awkward strings operations (`explode()`, `rtrim()`) that weren't necessary with the previous data structure.

You want a strong type-hint, create a `DatabaseTableName` class that will have an optional schema property, then you will have a strong type-hint.

If you want to stick with strings, I suggest you do the string manipulation in `DBA::quoteIdentifier` where you can split on periods and concatenate the result better than you're currently doing in `DBA::buildTableString`:
```php
public static function quoteIdentifier(string $identifier): string
{
  return implode(
    '.',
    array_map(
      function (string $identifier) { return '`' . str_replace('`', '``', $identifier) . '`'; },
      explode('.', $identifier)
    )
  );
}
```
And this way `DBA::buildTableString` becomes just:
```php
public static function buildTableString(array $tables): string
{
  return implode(',', array_map(['self', 'quoteIdentifier'], $tables));
}
```
Either way, please don't leave the codebase in the current state of your branch, it's the worst of both world.

Quix0r committed:
Use protected $this->logger instead

Quix0r commented:
With the old version `DBA::buildTableString()` would have no type-hint for the parameter, allowing all variables, including boolean and objects. `string` at least only allows strings.

Quix0r committed:
Rewrote to suggestions from @MrPetovan (I tried to maintain the old way of
['scheme' => 'table'] with my version).

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -218,12 +218,12 @@ public static function e(string $sql): bool
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 * @param array               $condition array of fields for condition
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -305,13 +304,12 @@ public static function insert($table, array $param, int $duplicate_mode = Databa
-	 * @param string|array $table Table name or array [schema => table]
-	 * @param array        $param parameter array
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -331,12 +329,11 @@ public static function lastInsertId(): int
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -385,13 +382,13 @@ public static function rollback(): bool
 	 *
-	 * @param string|array $table      Table name
-	 * @param array        $conditions Field condition(s)
+	 * @param string $table      Table name
```
Please update the documentation to reflect the type change (the possible schema prefix).

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -417,7 +414,7 @@ public static function delete($table, array $conditions, array $options = []): b
 	 * When you set $old_fields to "true" then $fields must contain all relevant fields!
 	 *
-	 * @param string|array  $table      Table name or array [schema => table]
+	 * @param string        $table      Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -450,27 +447,27 @@ public static function selectFirst($table, array $fields = [], array $condition
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -450,27 +447,27 @@ public static function selectFirst($table, array $fields = [], array $condition
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -487,17 +484,17 @@ public static function selectToArray($table, array $fields = [], array $conditio
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $condition array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBA.php:
```diff
@@ -524,37 +521,30 @@ public static function count($table, array $condition = [], array $params = []):
 	 * - [table1, table2, ...]
 	 * - [schema1 => table1, schema2 => table2, table3, ...]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/DBStructure.php:
```diff
@@ -1064,7 +1064,7 @@ function ($to, $from) {
 	 * @throws Exception
 	 */
-	public static function existsColumn($table, $columns = [])
+	public static function existsColumn(string $table, array $columns = []): bool
```
Please update the documentation to reflect the type change with the optional schema prefix.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -64,7 +64,7 @@ class Database
 	/** @var PDO|mysqli */
 	protected $connection;
-	protected $driver;
+	protected $driver = '';
```
Why the change to empty string?

Quix0r replied:
I have found a few occurrences:
- the getter `getDriver()` states `string` as return type or else `string` cannot be the returned type
- `self::assertContains()` asserts it against strings
- also all `switch()` usages are using only `string` cases.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -820,13 +820,13 @@ public function e(string $sql): bool
 	 *
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $condition array of fields for condition
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1003,14 +1003,14 @@ public function fetch($stmt)
-	 * @param string|array $table          Table name or array [schema => table]
-	 * @param array        $param          parameter array
-	 * @param int          $duplicate_mode What to do on a duplicated entry
+	 * @param string $table          Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1054,13 +1054,12 @@ public function insert($table, array $param, int $duplicate_mode = self::INSERT_
-	 * @param string|array $table Table name or array [schema => table]
-	 * @param array        $param parameter array
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1103,12 +1102,11 @@ public function lastInsertId(): int
 	 *
-	 * @param string|array $table Table name or array [schema => table]
-	 *
+	 * @param string $table Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1265,14 +1263,14 @@ public function rollback(): bool
 	 * @throws \Exception
 	 */
-	public function delete($table, array $conditions): bool
+	public function delete(string $table, array $conditions): bool
```
Please update the documentation to reflect the type change with the optional schema prefix.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1302,7 +1300,7 @@ public function delete($table, array $conditions): bool
 	 * When you set $old_fields to "true" then $fields must contain all relevant fields!
 	 *
-	 * @param string|array  $table      Table name or array [schema => table]
+	 * @param string        $table      Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1368,16 +1366,16 @@ public function update($table, array $fields, array $condition, $old_fields = []
-	 * @param array        $fields
-	 * @param array        $condition
-	 * @param array        $params
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1394,16 +1392,15 @@ public function selectFirst($table, array $fields = [], array $condition = [], a
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
-	 *
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1469,14 +1466,14 @@ private function escapeFields(array $fields, array $options): array
-	 * @param array        $fields    Array of selected fields, empty for all
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1509,9 +1506,9 @@ public function select($table, array $fields = [], array $condition = [], array
-	 * @param string|array $table     Table name or array [schema => table]
-	 * @param array        $condition Array of fields for condition
-	 * @param array        $params    Array of several parameters
+	 * @param string $table     Table name or array [schema => table]
```
Please update the documentation to reflect the type change.

MrPetovan commented on src/Database/Database.php:
```diff
@@ -1823,14 +1820,14 @@ public function escapeArray(&$arr, bool $add_quotation = false)
 	 * Replaces a string in the provided fields of the provided table
 	 *
-	 * @param string $table_name Table name
+	 * @param string $table  Table name
```
Please update the documentation to reflect the type change with the optional schema prefix.

Quix0r committed:
Fixed documentation

Quix0r committed:
Reverted and improved TODO:
------
@nupplaphil: There is no easy possibility to add a logger here, that's why there
isn't any yet and instead a placeholder.. This execution point is a critical
state during a testrun, and tbh I'd like to leave here no further logic (yet)
because I spent hours debugging cases, where transactions weren't fully closed
and strange/unpredictable errors occur (sometimes -mainly during debugging other
errors :) ...)
-----

MrPetovan commented on tests/Util/Database/StaticDatabase.php:
```diff
@@ -179,8 +179,7 @@ public static function statConnect(array $server)
 		} catch (PDOException $e) {
-			// Log exception
-			$this->logger::error('Cannot establish database connection', ['exception' => $e, 'host' => $server, 'dbname' => $db]);
+			/// @TODO Try to find a way to log this exception as it contains valueable information
```
Please add Nupplaphil's comment in the code for the next person who will try to refactor this.

Quix0r committed:
Added important comment from @nupplaphil@github.com
